pkg/odo/cli/component/exec.go

```diff
@@ -0,0 +1,113 @@
+package component
+
+import (
+	"fmt"
+	"github.com/openshift/odo/pkg/envinfo"
+	projectCmd "github.com/openshift/odo/pkg/odo/cli/project"
+	"github.com/openshift/odo/pkg/odo/genericclioptions"
+	odoutil "github.com/openshift/odo/pkg/odo/util"
+	"github.com/openshift/odo/pkg/odo/util/completion"
+	"github.com/openshift/odo/pkg/odo/util/experimental"
+	"github.com/openshift/odo/pkg/odo/util/pushtarget"
+	"github.com/openshift/odo/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	ktemplates "k8s.io/kubectl/pkg/util/templates"
+	"path/filepath"
+)
+
+// ExecRecommendedCommandName is the recommended exec command name
+const ExecRecommendedCommandName = "exec"
+
+var execExample = ktemplates.Examples(`  # Executes a command inside the component
+%[1]s
+`)
+
+// ExecOptions contains exec options
+type ExecOptions struct {
+	componentContext string
+	componentOptions *ComponentOptions
+	devfilePath      string
+	namespace        string
+
+	command []string
+}
+
+// NewExecOptions returns new instance of ExecOptions
+func NewExecOptions() *ExecOptions {
+	return &ExecOptions{
+		componentOptions: &ComponentOptions{},
+	}
+}
+
+// Complete completes exec args
+func (eo *ExecOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+	if cmd.ArgsLenAtDash() <= -1 {
+		return fmt.Errorf("no command was given for the exec command")
+	}
+
+	eo.command = args[cmd.ArgsLenAtDash():]
+
+	if len(eo.command) <= 0 {
+		return fmt.Errorf("no command was given for the exec command")
+	}
+
+	if len(eo.command) != len(args) {
+		return fmt.Errorf("no parameter is expected for the command")
+	}
+
+	eo.devfilePath = filepath.Join(eo.componentContext, eo.devfilePath)
+	// if experimental mode is enabled and devfile is present
+	if experimental.IsExperimentalModeEnabled() && util.CheckPathExists(eo.devfilePath) {
+		envInfo, err := envinfo.NewEnvSpecificInfo(eo.componentContext)
+		if err != nil {
+			return errors.Wrap(err, "unable to retrieve configuration information")
+		}
+		eo.componentOptions.Context = genericclioptions.NewDevfileContext(cmd)
+		eo.componentOptions.EnvSpecificInfo = envInfo
```

Comment:
this is being done as part of NewDevfileContext already https://github.com/openshift/odo/blob/a82f3e0256eace61ac66c344edff8b2d35d81331/pkg/odo/genericclioptions/context.go#L479